bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/BoschHttpClient.java

```diff
@@ -0,0 +1,139 @@
+package org.openhab.binding.boschshc.internal;
+
+import static org.eclipse.jetty.http.HttpMethod.GET;
+
+import java.nio.charset.StandardCharsets;
+import java.security.KeyStoreException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateEncodingException;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.util.StringContentProvider;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+public class BoschHttpClient extends HttpClient {
+
+    private static final int MAX_PAIR_TRIES = 12;
+    
+	private final Logger logger = LoggerFactory.getLogger(BoschHttpClient.class);
+
+	private String ipAddress;
+	private String systempassword;
+
+	public BoschHttpClient(String ipAddress, String systempassword, SslContextFactory sslContextFactory) {
+		super(sslContextFactory);
+		this.ipAddress = ipAddress;
+		this.systempassword = systempassword;
+	}
+
+	private String getCertFromSslContextFactory() throws KeyStoreException, CertificateEncodingException {
+		Certificate cert = this.getSslContextFactory().getKeyStore().getCertificate(BoschSslUtil.OSS_OPENHAB_BINDING);
+		return new String(Base64.getEncoder().encode(cert.getEncoded()));
+	}
+
+	public void checkAccessAndPairIfNecessary() throws InterruptedException {
+        // test if pairing is needed
+   		int counter = 0;
+       	boolean accessPossible = isAccessPossible();
+       	while (!accessPossible) {
+			// sleep some seconds after every try, except the first one
+   			if (counter > 0) {
+   				logger.info("Last Pairing failed, starting retry number {}/{} in few seconds", counter, MAX_PAIR_TRIES);
+       			Thread.sleep(15000);
+   			} else {
+   				logger.info("Pairing needed, because access to Bosch SmartHomeController not possible.");
+   			}
+   			// Timeout after max tries with an exception that pairing failed
+   			if (counter >= MAX_PAIR_TRIES)  {
+   				// all tries during the last MAX_PAIR_TRIES*1500 milliseconds failed
+   				throw new IllegalStateException("Pairing abourted, because it failed too many times!");
+   			}
+   			// try to pair if no access is possible 
+   			counter++;
+   			doPairing();
+   			// check access
+   			accessPossible = isAccessPossible();
+   		} 
```

Comment:
I guess we should mark the binding offline somewhere if the pairing fails. I don't think I have seen this in your code yet? Perhaps this function could return a boolean, and we could use that to set the state of the binding so that OH knows whether it's online? There, we can also set a reason for the incorrect initialization as a string that will be displayed to the user.

Reply:
That's great solution if the pairing wasn't possible! We can avoid an endless loop *and* give after the duration a feedback via the "offline" state.
Only if pairing and access are OK, we change to "online".

Reply:
The Hue Binding Code seems to do a pairing whenever a call fails with an exception, see [withReAuthentication](https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java#L920) and callers and compare with [onNotAuthenticated()](https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java#L717)
There are no number of tries, no timeouts, ...

The bridge is set "offline" with a nice  description if pairing is necessary, see ![2020-06-27 16_34_33-Paper UI](https://user-images.githubusercontent.com/1284770/85924748-740e8280-b894-11ea-8d12-af9036bd2a45.jpg)

Reply:
Implemented status update works now if all pairings failed
![2020-06-27 20_01_48-Paper UI](https://user-images.githubusercontent.com/1284770/85928930-25bbac80-b8b1-11ea-850a-6795e9855da8.jpg)

Reply:
Very cool!